### ISSUE TITLE
Question about Reppoints head's inconsistent behaviors during training and inference

### ISSUE THREAD

**Issue by sandylaker** [open-mmlab/mmdetection#3715]
Hi, I have a minor question regarding the `RepPointsHead`.

At training time, the pseudo boxes from the initial stage serve as anchor boxes for the refine stage. 
```python3
for i_img, center in enumerate(center_list):
            bbox = []
            for i_lvl in range(len(pts_preds_refine)):
                bbox_preds_init = self.points2bbox(
                    pts_preds_init[i_lvl].detach())
                bbox_shift = bbox_preds_init * self.point_strides[i_lvl]
                bbox_center = torch.cat(
                    [center[i_lvl][:, :2], center[i_lvl][:, :2]], dim=1)
                bbox.append(bbox_center +
                            bbox_shift[i_img].permute(1, 2, 0).reshape(-1, 4))
            bbox_list.append(bbox)
        cls_reg_targets_refine = self.get_targets(
            bbox_list,
            valid_flag_list,
            gt_bboxes,
            img_metas,
            gt_bboxes_ignore_list=gt_bboxes_ignore,
            gt_labels_list=gt_labels,
            stage='refine',
            label_channels=label_channels)
```
However, at inference stage, it directly uses the grid points to decode the predictions into bounding boxes, instead of using the pseudo boxes.
```python3
def get_bboxes(self,
                   cls_scores,
                   pts_preds_init,
                   pts_preds_refine,
                   img_metas,
                   cfg=None,
                   rescale=False,
                   nms=True):
        assert len(cls_scores) == len(pts_preds_refine)
        bbox_preds_refine = [
            self.points2bbox(pts_pred_refine)
            for pts_pred_refine in pts_preds_refine
        ]
        num_levels = len(cls_scores)
        mlvl_points = [
            self.point_generators[i].grid_points(cls_scores[i].size()[-2:],
                                                 self.point_strides[i])
            for i in range(num_levels)
        ]
        result_list = []
        for img_id in range(len(img_metas)):
            cls_score_list = [
                cls_scores[i][img_id].detach() for i in range(num_levels)
            ]
            bbox_pred_list = [
                bbox_preds_refine[i][img_id].detach()
                for i in range(num_levels)
            ]
            img_shape = img_metas[img_id]['img_shape']
            scale_factor = img_metas[img_id]['scale_factor']
            proposals = self._get_bboxes_single(cls_score_list, bbox_pred_list,
                                                mlvl_points, img_shape,
                                                scale_factor, cfg, rescale,
                                                nms)
            result_list.append(proposals)
        return result_list
```
This becomes inconsistent with the procedures at training time. According to the equation (5) and Figure 3. from the RepPoints paper, the predictions at refine stage should be added onto the pseudo boxes. Thus, the `loss` function in the implementation matches the description in the paper, while `get_bboxes` function does not. Could anyone explain it for me? Any suggestion would be highly appreciated.

Thanks.

**Comment by sandylaker** [comment]
Sorry, I missed the fact that in `self.get_targets`, it only assigned gts to each pseudo boxes, but did not encode them like in `AnchorHead`. Therefore, `pts_out_refine = pts_out_refine + pts_out_init.detach()`  is actually correct, it predicts the offsets between grid points and gt boxes, rather than offsets between pseudo boxes and gt boxes.
